Add documentation create/update community providers (#15061)
mistercrunch/Airflow: docs/apache-airflow-providers/howto/create-update-providers.rst

 .. Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

 ..   http://www.apache.org/licenses/LICENSE-2.0

 .. Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

Community Providers
===================

.. contents:: :local:

How-to creating a new community provider
----------------------------------------

This document gathers the necessary steps to create a new community provider and also guidelines for updating

> the existing ones. You should be aware that providers may have distinct distinctions that may not be covered in

Reviewer: ```suggestion
the existing ones. You should be aware that providers may have distinctions that may not be covered in
```

if you didn't set up your local environment yet. We recommend using the set up through ``breeze``. This way you
will easily be able to have an environment more similar to the one executed by the validation code of Github CI
workflow.

  .. code-block:: bash

      ./breeze

Using the code above you will set up Docker containers. These containers your local code to internal volumes.
In this way, the changes made in your IDE are already applied to the code inside the container and tests can
be carried out quickly.

Initial Code and Unit Tests
^^^^^^^^^^^^^^^^^^^^^^^^^^^

Most likely you have developed a version of the provider using some local customization and now you need to
transfer this code to the Airflow project. Below is described all the initial code structure that
the provider may need. Understand that not all providers will need all the components described in this structure.
If you still have doubts about building your provider, we recommend that you read the initial provider guide and
open a issue on Github so the community can help you.

  .. code-block:: bash

      airflow/

>       ├── providers/new_provider/

Reviewer: Maybe we should change `new_provider` into `<NEW_PROVIDER>` all over the places to indicate that it should be replaced with the real name?

- add your provider to ``ALL_PROVIDERS`` list;
- add your provider into ``CONNECTIONS_LIST`` if your provider create a new connection type.

Integration tests
^^^^^^^^^^^^^^^^^

[to do]
See `Airflow Integration Tests <https://github.com/apache/airflow/blob/master/TESTING.rst#airflow-integration-tests>`_

Documentation
^^^^^^^^^^^^^

An important part of building a new provider is the documentation.

  .. code-block:: bash

      airflow/
      ├── INSTALL
      ├── CONTRIBUTING.rst
      ├── setup.py
      ├── docs/
      │   ├── spelling_wordlist.txt
      │   ├── apache-airflow/
      │   │   └── extra-packages-ref.rst
      │   ├── integration-logos/new_provider/
      │   │   └── New_Provider.png
      │   └── apache-airflow-providers-new_provider/
      │       ├── index.rst
      │       ├── commits.rst
      │       ├── connections.rst
      │       └── operators/
      │           └── new_provider.rst
      └── providers/
          ├── dependencies.json
          └── new_provider/
              ├── provider.yaml
              └── CHANGELOG.rst

The ``airflow/providers/dependencies.json`` is automatically updated by pre-commit.

> There is a chance that your provider's name is not an common English word.

Reviewer: ```suggestion
There is a chance that your provider's name is not a common English word.
```

> In this case is necessary to add it to the file ``docs/spelling_wordlist.txt``. This file begin with capitalize words and

Reviewer: ```suggestion
In this case is necessary to add it to the file ``docs/spelling_wordlist.txt``. This file begin with capitalized words and
```

lowercase in the second block.

  .. code-block:: bash

    Namespace
    Neo4j
    Nextdoor
    New_Provider (new line)
    Nones
    NotFound
    Nullable
    ...
    neo4j
    neq
    networkUri
    new_provider (new line)
    nginx
    nobr
    nodash

Add your provider dependencies into **PROVIDER_REQUIREMENTS** variable in ``setup.py``. If your provider doesn't have
any dependency add a empty list.

  .. code-block:: python

      PROVIDERS_REQUIREMENTS: Dict[str, List[str]] = {
          ...
          'microsoft.winrm': winrm,
          'mongo': mongo,
          'mysql': mysql,
          'neo4j': neo4j,
          'new_provider': [],
          'odbc': odbc,
          ...
          }

In the ``CONTRIBUTING.rst`` adds:

- your provider name in the list in the **Extras** section
- your provider dependencies in the **Provider Packages** section table, only if your provider has external dependencies.

In the ``INSTALL`` file adds:
- your provider to the **Extras** section list.

> 

Reviewer: This will happen automatcally via pre-commit. Might be a good idea to add step above about installing pre-commit and adding `pre-commit install`

In the ``docs/apache-airflow-providers-new_provider/connections.rst``:

> - add information how to connect to your providers.

Reviewer: ```suggestion
- add information how to configure connection for your provider.
```

In the ``docs/apache-airflow-providers-new_provider/operators/new_provider.rst``:

- add information how to use the Operator. It's important to add examples and additional information if your Operator has extra-parameters.

  .. code-block:: RST

      .. _howto/operator:NewProviderOperator:

      NewProviderOperator
      ===================

      Use the :class:`~airflow.providers.new_provider.operators.NewProviderOperator` to do something
      amazing with Airflow!

      Using the Operator
      ^^^^^^^^^^^^^^^^^^

      The NewProviderOperator requires a ``connection_id`` and this other awesome parameter.
      You can see an example below:

      .. exampleinclude:: /../../airflow/providers/new_provider/example_dags/example_new_provider.py
          :language: python
          :start-after: [START howto_operator_new_provider]
          :end-before: [END howto_operator_new_provider]

> In the ``docs/apache-airflow-providers-new_provider/commits.rst``:

Reviewer: This is not needed . It will be added automatically when providers gets released.

- add the initial information of your providers. This file is updated automatically by Airflow. Below is shown an example.

  .. code-block:: RST

      package apache-airflow-providers-new_provider
      ---------------------------------------------

      `New_Provider <https://example.io/>`__

      This is detailed commit list of changes for versions provider package: ``new_provider``.
      For high-level changelog, see :doc:`package information including changelog <index>`.

In the ``docs/apache-airflow-providers-new_provider/index.rst``:

- add all information of the purpose of your provider. It is recommended to check with another provider to help you complete this document as best as possible.

> In the ``/airflow/providers/new_provider/CHANGELOG`` add the initial information. Providers start with 1.0.0 version.

Reviewer: Same here. No need to add it, it will be added automatically during release.

In the ``airflow/providers/new_provider/provider.yaml`` add information of your provider:

  .. code-block:: yaml

      package-name: apache-airflow-providers-new_provider
      name: New_Provider
      description: |
        `New_Provider <https://example.io/>`__
      versions:
        - 1.0.0

      integrations:
        - integration-name: New_Provider
          external-doc-url: https://www.example.io/
          logo: /integration-logos/new_provider/New_Provider.png
          how-to-guide:
            - /docs/apache-airflow-providers-new_provider/operators/new_provider.rst
          tags: [service]

      operators:
        - integration-name: New_Provider
          python-modules:
            - airflow.providers.new_provider.operators.new_provider

      hooks:
        - integration-name: New_Provider
          python-modules:
            - airflow.providers.new_provider.hooks.new_provider

      sensors:
        - integration-name: New_Provider
          python-modules:
            - airflow.providers.new_provider.sensors.new_provider

      hook-class-names:
        - airflow.providers.new_provider.hooks.new_provider.NewProviderHook

> After changing and creating these files you can build the documentation locally. The two commands below will

Reviewer: ```suggestion
You only need to add `hook-class-names` in case you have some hooks that have customized UI behavior. 

After changing and creating these files you can build the documentation locally. The two commands below will
```

Also reference to http://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html#custom-connection-types would be nice.

serve to accomplish this. The first will build your provider's documentation. The second will ensure that the
main Airflow documentation that involves some steps with the providers is also working.

  .. code-block:: bash

    ./breeze build-docs -- --package-filter apache-airflow-providers-new_provider
    ./breeze build-docs -- --package-filter apache-airflow

How-to Update a community provider
----------------------------------

> [to do]

Reviewer: This is the work of release manager. It is described in https://github.com/apache/airflow/blob/master/dev/README_RELEASE_PROVIDER_PACKAGES.md#provider-packages-versioning
